xarray/backends/rasterio_.py

```diff
@@ -5,127 +5,109 @@
+def rasterio_to_dataarray(filename):
+    """Open a file with rasterio.
+
+    This should work with any file that rasterio can open (most often: 
+    geoTIFF). The x and y coordinates are generated automatically from the 
+    file's geoinformation.
     """
-    def __init__(self, filename, mode='r'):
-
-        with rasterio.Env():
-            self.ds = rasterio.open(filename, mode=mode)
-
-        # Get coords
-        nx, ny = self.ds.width, self.ds.height
-        dx, dy = self.ds.res[0], -self.ds.res[1]
-        x0 = self.ds.bounds.right if dx < 0 else self.ds.bounds.left
-        y0 = self.ds.bounds.top if dy < 0 else self.ds.bounds.bottom
-        x = np.linspace(start=x0, num=nx, stop=(x0 + (nx - 1) * dx))
-        y = np.linspace(start=y0, num=ny, stop=(y0 + (ny - 1) * dy))
-
-        self._vars = OrderedDict()
-        self._vars['y'] = Variable(('y',), y)
-        self._vars['x'] = Variable(('x',), x)
-
-        # Get dims
-        if self.ds.count >= 1:
-            self.dims = ('band', 'y', 'x')
-            self._vars['band'] = Variable(('band',),
-                                          np.atleast_1d(self.ds.indexes))
-        else:
-            raise ValueError('Unknown dims')
-
-        self._attrs = OrderedDict()
-        with suppress(AttributeError):
-            for attr_name in ['crs']:
-                self._attrs[attr_name] = getattr(self.ds, attr_name)
-
-        # Get data
-        self._vars[_rio_varname] = self.open_store_variable(_rio_varname)
-
-    def open_store_variable(self, var):
-        if var != _rio_varname:
-            raise ValueError('Rasterio variables are named %s' % _rio_varname)
-        data = indexing.LazilyIndexedArray(RasterioArrayWrapper(self.ds))
-        return Variable(self.dims, data, self._attrs)
-
-    def get_variables(self):
-        return FrozenOrderedDict(self._vars)
-
-    def get_attrs(self):
-        return Frozen(self._attrs)
-
-    def get_dimensions(self):
-        return Frozen(self.dims)
-
-    def close(self):
-        self.ds.close()
+
+    riods = rasterio.open(filename, mode='r')
+
+    # Get geo coords
+    nx, ny = riods.width, riods.height
+    dx, dy = riods.res[0], -riods.res[1]
+    x0 = riods.bounds.right if dx < 0 else riods.bounds.left
+    y0 = riods.bounds.top if dy < 0 else riods.bounds.bottom
+    x = np.linspace(start=x0, num=nx, stop=(x0 + (nx - 1) * dx))
+    y = np.linspace(start=y0, num=ny, stop=(y0 + (ny - 1) * dy))
+
+    # Get bands
+    if riods.count < 1:
+        raise ValueError('Unknown dims')
+    bands = np.asarray(riods.indexes)
+
+    # Attributes
+    attrs = {}
+    if hasattr(riods, 'crs'):
+        # CRS is a dict-like object specific to rasterio
+        # We convert it back to a PROJ4 string using rasterio itself
+        attrs['crs'] = riods.crs.to_string()
+    # Maybe we'd like to parse other attributes here (for later)
+
+    data = indexing.LazilyIndexedArray(RasterioArrayWrapper(riods))
```

Comment:
Implementing ``cache`` went well (appart from this [bug](https://github.com/pydata/xarray/issues/1429).

For chunking I have a question: ``DataArray.chunk()`` [doesn't have](http://xarray.pydata.org/en/stable/generated/xarray.DataArray.chunk.html#xarray.DataArray.chunk) the ``name_prefix``, ``token`` and ``lock`` keywords. Any reason for this?